pywemo/subscribe.py

```diff
@@ -35,32 +73,124 @@ def get_ip_address(host='1.2.3.4'):
+      to update the BinaryState of the virtual device. This doesn't actually
+      update any state, rather the virtual device then delivers the event
+      notification to any event listeners configured to receive events from the
+      pywemo SubscriptionRegistry. The event type for a long press action is
+      EVENT_TYPE_LONG_PRESS.
+    """
+
+    def do_NOTIFY(self):  # pylint: disable=invalid-name
         """Handle subscription responses received from devices."""
         sender_ip, _ = self.client_address
         outer = self.server.outer
         device = outer.devices.get(sender_ip)
-        content_len = int(self.headers.get('content-length', 0))
-        data = self.rfile.read(content_len)
         if device is None:
             LOG.warning('Received event for unregistered device %s', sender_ip)
         else:
-            # trim garbage from end, if any
-            data = data.decode("UTF-8").split("\n\n")[0]
-            doc = cElementTree.fromstring(data)
+            doc = self._get_xml_from_http_body()
             for propnode in doc.findall('./{0}property'.format(NS)):
                 for property_ in list(propnode):
                     text = property_.text
                     outer.event(device, property_.tag, text)
 
-        self.send_response(200)
-        self.send_header('Content-Type', 'text/html')
-        self.send_header('Content-Length', len(SUCCESS))
+        self._send_response(200, RESPONSE_SUCCESS)
+
+    def do_GET(self):  # pylint: disable=invalid-name
+        """Handle GET requests for a Virtual WeMo device."""
+        if self.path.endswith("/setup.xml"):
+            self._send_response(
+                200, VIRTUAL_SETUP_XML, content_type="text/xml"
+            )
+        else:
+            self._send_response(404, RESPONSE_NOT_FOUND)
+
+    def do_POST(self):  # pylint: disable=invalid-name
+        """Handle POST requests for a Virtual WeMo device."""
+        if self.path.endswith("/upnp/control/basicevent1"):
+            sender_ip, _ = self.client_address
+            outer = self.server.outer
+            device = outer.devices.get(sender_ip)
+            if device is None:
+                LOG.warning(
+                    'Received event for unregistered device %s', sender_ip
+                )
+            else:
+                doc = self._get_xml_from_http_body()
+                binary_state = doc.find('.//BinaryState')
+                if binary_state is not None:
+                    text = binary_state.text
+                    outer.event(device, EVENT_TYPE_LONG_PRESS, text)
+            self._send_response(200, RESPONSE_SUCCESS)
+        else:
+            self._send_response(404, RESPONSE_NOT_FOUND)
+
+    def do_SUBSCRIBE(self):  # pylint: disable=invalid-name
+        """Handle SUBSCRIBE requests for a Virtual WeMo device."""
+        if self.path.endswith("/upnp/event/basicevent1"):
+            self.send_response(200)
```

Comment:
Not particularly important, but you could modify the `_send_response` function to accept an optional dictionary (which are ordered since python 3.7) of data to send after the content-length.  Then you could reuse that function here by using an empty string for the body and passing something like `extra_headers={'TIMEOUT': 'Second-1801', 'SID': 'uuid:virtual-device-uuid'}` for that new keyword.

Reply:
I could go either way (keep it as-is or modify `_send_response` to take the headers as a parameter). But Wemo doesn't respond with a content-type header, so there might end up being a few more edge cases to handle in `_send_response` if we went that route. If it's okay, let's keep it as-is for now. That way the special-casing is in just the `do_SUBSCRIBE` method. 

```
HTTP/1.1 200 OK
DATE: Mon, 04 Jan 2021 22:26:58 GMT
SERVER: Unspecified, UPnP/1.0, Unspecified
CONTENT-LENGTH: 0
X-User-Agent: redsonic
SID: uuid:fa79d9d4-1dd1-11b2-9603-8f01388fd38f
TIMEOUT: Second-1801

```

Reply:
Sounds good.